### PR TITLE
Fix regression in names of the template-based projects

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

[ci no changelog needed]

A regression was introduced by the latest Welcome Screen changes. Projects created by opening example from Welcome screen have names "Unnamed" instead of ie. "Stargazers".

This PR fixes the bug.

### Important Notes

The project manager requires project names to start with an uppercase letter, but templates names are all-lowercased actually. This requires capitalizing the first letter of the template name before sending it as the project name.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: app/gui/src/controller/ide/desktop.rs [hunk 1]
```diff
@@ -164,3 +161,27 @@ impl ManagingProjectAPI for Handle {
...
 }
+
+/// Select a new name for the project in a form of <suggested_name>_N, where N is a unique sequence
+/// number.
+fn choose_new_project_name(
+    mut existing_names: impl Iterator<Item = String>,
+    suggested_name: &str,
+) -> String {
+    let with_suffix = (1..).map(|i| format!("{}_{}", suggested_name, i));
+    let mut candidates = std::iter::once(suggested_name.to_owned()).chain(with_suffix);
```
Comment (mwu-tow) [inline app/gui/src/controller/ide/desktop.rs]: I don't think this should work. Was it tested? 
`Foo_1` is not a valid project name, as each segment must start with an uppercase letter. `1` is not one.

File: app/gui/src/controller/ide/desktop.rs [hunk 2]
```diff
@@ -164,3 +161,27 @@ impl ManagingProjectAPI for Handle {
         .boxed_local()
     }
 }
+
+/// Select a new name for the project in a form of <suggested_name>_N, where N is a unique sequence
+/// number.
+fn choose_new_project_name(
+    mut existing_names: impl Iterator<Item = String>,
+    suggested_name: &str,
+) -> String {
```
Comment (mwu-tow) [inline app/gui/src/controller/ide/desktop.rs]: `ReferentName` should be used for project name.

File: app/gui/src/controller/ide/desktop.rs [hunk 3]
```diff
@@ -164,3 +161,27 @@ impl ManagingProjectAPI for Handle {
...
+    let with_suffix = (1..).map(|i| format!("{}_{}", suggested_name, i));
+    let mut candidates = std::iter::once(suggested_name.to_owned()).chain(with_suffix);
+    // The iterator have no end, so we can safely unwrap.
+    let name = candidates.find(|c| !existing_names.contains(c)).unwrap();
+    capitalize_first(&name)
```
Comment (mwu-tow) [inline app/gui/src/controller/ide/desktop.rs]: This should capitalize each segment first letter, and fix the string if segments does not start with a letter.

File: app/gui/src/controller/ide/desktop.rs [hunk 4]
```diff
@@ -164,3 +161,27 @@ impl ManagingProjectAPI for Handle {
...
+}
+
+/// Capitalize the first letter of the string.
+///
+/// CC-BY-SA 4.0 https://stackoverflow.com/a/38406885
```
Comment (mwu-tow) [inline app/gui/src/controller/ide/desktop.rs]: IANAL, but I don't think this counts as a proper attribution.
On the other hand I don't think that such trivial snippet should be even treated as a work in sense of copyright law.

File: app/gui/src/controller/ide/desktop.rs (hunk 3 above)
Comment (mwu-tow) [reply app/gui/src/controller/ide/desktop.rs]: Also, such normalization could be a function in `ReferentName`.

**Review by mwu-tow — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/ide/desktop.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: > Foo_1 is not a valid project name, as each segment must start with an uppercase letter. 1 is not one.

`Foo_1` is actually a totally valid project name. Please see [`ReferentName`](https://github.com/enso-org/enso/blob/f03640b93a242237f78e6095c9d19cc16f4cb8ce/app/gui/controller/double-representation/src/identifier.rs#L213) documentation.

Comment (mwu-tow) [reply app/gui/src/controller/ide/desktop.rs]: The `ReferentName` is known to be incomplete when it comes to validation: https://github.com/enso-org/enso/issues/2797
However, you are right about `Foo_1` being valid, non-first words within an identifier are allowed to be numbers.
See https://github.com/enso-org/enso/blob/develop/docs/syntax/naming.md#naming-constructs for the official source of truth about identifiers.

File: app/gui/src/controller/ide/desktop.rs (hunk 2 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: I think `ProjectName` suits the situation much better. I have also changed the Project Manager API definition to use `ProjectName` as well.

File: app/gui/src/controller/ide/desktop.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: There is no need for such capitalizations, but you're right, the `ProjectName` constructor now does the capitalization of the first letter.

File: app/gui/src/controller/ide/desktop.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: It's an extremely complicated legal stuff. I think this is considered a proper attribution, and this code sample will probably fall under "fair use" conditions. However, I changed it significantly while refactoring the code, and removed attribution, therefore. It takes too much effort to copy code from SO :(

File: app/gui/src/controller/ide/desktop.rs (hunk 2 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: To clarify, `ReferentName` seems to be `controller` entity, but here were are interacting with Project Manager, which apparently uses `ProjectName` (well, it should).

**Review by wdanilo — APPROVED** [review]

**Review by mwu-tow — APPROVED** [review]

**Pull request merged**